_posts/2023-03-02-dafny-4-released.markdown

```diff
@@ -0,0 +1,97 @@
+
+### Support for Testing
+Dafny provides experimental support for writing unit tests akin to popular unit testing and mocking frameworks that are available for mainstream programming languages. For example, you can now use the `:test` annotation to tag your unit tests in Dafny. Then, the Dafny to C# compiler seamlessly compiles such Dafny tests into XUnit tests that can be executed on the compiled Dafny program. Similarly, there is now a `:synthesize` annotation for mocking.
+
+Dafny supports an experimental feature for automatic test generation for statement-level code coverage. The tests contain runtime checks compiled directly from (non-ghost) postconditions. Compared to prior versions, Dafny 4.0 can generate tests both for methods and functions with support for all basic types, collections, algebraic datatypes, class types, and type parameters as well as limited support for traits, function types, and tuples. To invoke test generation using the new command line interface, execute `dafny generate-tests`. For more information please read this [README](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyTestGeneration/README.md).
+
+### New language features
+Compared to Dafny 3.0, Dafny 4.0 includes the following language improvements:
+* `for` loops,
+* `continue` statements,
+* parameter passing via name bindings,
+* parameter default values,
+* `function by method` declarations,
+* ghost tuples,
+* ghost allocation of class instances,
+* ghost datatype constructors,
+* `<-` syntax for quantified variables,
+* disjunctive `case ... | ... =>` patterns,
+* opaque constants, and
+* command-line arguments access from `Main`.
+
+### Documentation
+The language reference manual has been updated for all features in version 4.0. We have also improved the wording of error messages, and we added an [error catalog](https://dafny.org/latest/HowToFAQ/Errors) where you can get more information and helpful hints about error messages.
+
+## Backward compatibility considerations
+
+### Functions compiled by default
+If you’ve used Dafny before, the first change you’ll notice is how functions are declared in Dafny 4.0. Dafny distinguishes *methods*, which consist of code with possible side effects, from *functions*, which, unlike the “functions” in C or JavaScript, are like mathematical functions without side effects—you evaluate a function twice in Dafny and you’ll get the same value. Functions in Dafny were first used mostly in specifications, so they were by default *ghost*, which means the compiler erases them. However, because so many Dafny programs find functions to be useful also in compiled code, we have simplified how functions are declared. In Dafny 4.0, use the keyword `function` to declare a compiled function and use the keywords `ghost function` to declare a ghost function.
```

Comment:
```suggestion
If you’ve used Dafny before, the first change you’ll notice is how functions are declared in Dafny 4.0. Dafny distinguishes *methods*, which consist of code with possible side effects, from *functions*, which, unlike the “functions” in C or JavaScript, are like mathematical functions without side effects — you evaluate a function twice in Dafny and you’ll get the same value. Functions in Dafny were first used mostly in specifications, so they were by default *ghost*, which means the compiler erases them. However, because so many Dafny programs find functions to be useful also in compiled code, we have simplified how functions are declared. In Dafny 4.0, use the keyword `function` to declare a compiled function and use the keywords `ghost function` to declare a ghost function.
```